File: Muc_Do_4_Rat_Kho/Bai16_HaiDoanKhongGiao/Solution.py
import sys
def solve():
    try:
        sys.stdin = open('BAI16.INP', 'r')
        sys.stdout = open('BAI16.OUT', 'w')
    except:
        pass
    data = sys.stdin.read().split()
    if not data: return
    n = int(data[0])
    arr = [int(x) for x in data[1:1+n]]
    
    pref_max = [-float('inf')] * n
    cur = 0
    mx = -float('inf')
    for i in range(n):
        cur = max(arr[i], cur + arr[i])
        mx = max(mx, cur)
        pref_max[i] = mx
        
    suff_max = [-float('inf')] * n
    cur = 0
    mx = -float('inf')
    for i in range(n - 1, -1, -1):
        cur = max(arr[i], cur + arr[i])
        mx = max(mx, cur)
        suff_max[i] = mx
        
    ans = -float('inf')
    for i in range(n - 1):
        ans = max(ans, pref_max[i] + suff_max[i + 1])
    print(ans)
solve()

File: Muc_Do_4_Rat_Kho/Bai16_HaiDoanKhongGiao/Solution.cpp
#include <bits/stdc++.h>
using namespace std;
// Bai 16: HaiDoanKhongGiao (Max sum of 2 non-overlapping subarrays)
int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    if (fopen("BAI16.INP", "r")) {
        freopen("BAI16.INP", "r", stdin);
        freopen("BAI16.OUT", "w", stdout);
    }
    int n;
    if (!(cin >> n)) return 0;
    vector<long long> a(n);
    for (int i = 0; i < n; ++i) cin >> a[i];
    
    vector<long long> pref_max(n, -1e18);
    long long cur = 0, mx = -1e18;
    for (int i = 0; i < n; ++i) {
        cur = max(a[i], cur + a[i]);
        mx = max(mx, cur);
        pref_max[i] = mx;
    }
    
    vector<long long> suff_max(n, -1e18);
    cur = 0; mx = -1e18;
    for (int i = n - 1; i >= 0; --i) {
        cur = max(a[i], cur + a[i]);
        mx = max(mx, cur);
        suff_max[i] = mx;
    }
    
    long long ans = -1e18;
    for (int i = 0; i < n - 1; ++i) {
        ans = max(ans, pref_max[i] + suff_max[i + 1]);
    }
    cout << ans << "\n";
    return 0;
}
